### ISSUE TITLE
Using other optimizer

### ISSUE THREAD

**Issue by AlexanderGeng** [puneet29/StyleTransferApp#3]
Hello,

@puneet29 first of all your code and your ReadMe is very good.
In my master thesis i want to compare different optimizer in the training step.

In Pytorch there are several precoded optimizer.
Most of them work but up to now the optimizer SGD, ASGD, SGD(nesterov variant) don't work in my case.

If i want to use them, after training i get only a black image if i test the model with stylize.py.

I have also looked at the (total_style_loss+total_content_loss)/(batch_id+1) value.
The first one is 7051254.9375, which looks correct but then it explodes up to a value of 2.1273383372139363e+21 in my case and after that there are only "nan" outputs.

Do you have an idea why this happened?

Best regards
Alex

**Comment by puneet29** [comment]
Hey @Alex-259,

Thank you for taking the time to go through this repository. I will have to test for these optimizers. Can you please also share other variables that you are using while training?

Thanks and Regards,
Puneet

**Comment by AlexanderGeng** [comment]
Hey @puneet29 

after i recognized, that when i use the models (ASGD, SGD, SGD nesterov) with the stilize.py command i tried to change the SGD optimizer that it gives better results.

What i have done so far is the following:
I replaced in the train.py file the optimizer from Adam (which work fine) to the other variants of optimizer.
Adadelta, Adagrad, Adam amsgrad variant, Adam, Adamax, Rmsprop, Rprop worked only by replacing the optimizer.
So i wanted to do the same thing only replace the optimizer with SGD, ASGD, or SGD nesterov variant.
Unfortunately the loss ((total_style_loss+total_content_loss)/(batch_id+1)) explodes after three steps up to "nan".

I also tried to change the learning rate in the SGD case but also lr=0,1, lr=0,0001 and lr=0,00000001 didn't solve the problem.
After some research in the web somebody adivise to use the MSELoss with reduction='sum' or with reduction='elementwise_mean' as input.
The Losses don't explode in the way as before but the values don't change as you can see in the following saved .txt file
[listfile_sgd_reduction.txt](https://github.com/puneet29/StyleTransferApp/files/4343900/listfile_sgd_reduction.txt)
 
And when i tested the trained model with the stylize.py command i get still black images.

I used as you mentioned in the ReadMe the Coco Training images Dataset, as epoch=2, cuda=1 and as style-image the following image that i have downloaded from the internet and also successfully used with the other optimizer.
![farbig](https://user-images.githubusercontent.com/60505803/76866831-f97da200-6864-11ea-98f6-9f61fbf3a2de.jpg)

**Comment by puneet29** [comment]
I am not quite sure why this would happen but you can try altering the values for _style and content weight_. Maybe that can show some improvements?

**Comment by AlexanderGeng** [comment]
@puneet29  Thanks for your quick answer.
I have tried several values(both 1, one of them high, one of them low, both high, both low,...) for the content and style weight but in most of the cases i also get nan as output. The only possibility was both with the value 10 then i get no nan value, but the other loss values only change a little bit and as result image from stylize.py i get unfortunately also a black image. Here is my output of the training steps. I added a detection of nan values

![Output](https://user-images.githubusercontent.com/60505803/76941890-16fe4a80-68fd-11ea-9082-92f6c562ecc2.png)
